test/python/circuit/test_controlled_gate.py

```diff
@@ -290,8 +292,124 @@ def test_multi_control_u1(self):
+
+        # iterate over all possible combinations of control qubits
+        allsubsets = list(itertools.chain(*[itertools.combinations(range(num_controls), ni)
+                                            for ni in range(num_controls + 1)]))
+        for subset in allsubsets:
+            control_int = 0
+            lam = 0.3165354 * pi
+            qc = QuantumCircuit(q_target, q_controls)
+            for idx in subset:
+                control_int += 2**idx
+                qc.x(q_controls[idx])
+
+            qc.h(q_target[0])
+            qc.mcu1(lam, q_controls, q_target[0])
+            qc.h(q_target[0])
+
+            for idx in subset:
+                qc.x(q_controls[idx])
+
+            backend = BasicAer.get_backend('unitary_simulator')
+            mat_mcu = execute(qc, backend).result().get_unitary(qc)
+
+            dim = 2**(num_controls + 1)
+            pos = dim - 2 * (control_int + 1)
+            mat_groundtruth = np.eye(dim, dtype=complex)
+            fac = np.exp(1j * lam)
+            mat_groundtruth[pos:pos + 2, pos:pos + 2] = [[(1 + fac) / 2, (1 - fac) / 2],
+                                                         [(1 - fac) / 2, (1 + fac) / 2]]
+            self.assertTrue(np.allclose(mat_mcu, mat_groundtruth))
+
+    @unpack
+    @data(*itertools.product(
+        [1, 2, 3], ['basic']
+    ))
+    def test_multi_control_toffoli_matrix_clean_ancillas(self, num_controls, mode):
+        """Test the multi-control Toffoli gate with clean ancillas."""
+
+        # set up circuit
+        q_controls = QuantumRegister(num_controls)
+        q_target = QuantumRegister(1)
+        qc = QuantumCircuit(q_target, q_controls)
+
+        # add ancillas if necessary
+        num_ancillas = 0 if num_controls <= 2 else num_controls - 2
+        q_ancillas = None
+        if num_ancillas > 0:
+            q_ancillas = QuantumRegister(num_ancillas)
+            qc.add_register(q_ancillas)
+
+        # apply hadamard on control qubits and toffoli gate
+        qc.h(q_controls)
+        qc.mct(q_controls, q_target[0], q_ancillas, mode=mode)
+
+        # execute the circuit and obtain statevector result
+        backend = BasicAer.get_backend('statevector_simulator')
+        vec_mct = execute(qc, backend).result().get_statevector(qc)
+
+        # compare to expectation
+        mat = np.eye(2 ** (num_controls + 1))
+        mat[-2:, -2:] = [[0, 1], [1, 0]]
+        if num_ancillas > 0:
+            mat = np.kron(np.eye(2 ** num_ancillas), mat)
+
+        vec_groundtruth = mat @ np.kron(np.kron(
+            np.array([1] + [0] * (2 ** num_ancillas - 1)),
+            [1 / 2 ** (num_controls / 2)] * 2 ** num_controls), [1, 0])
+
+        s_f = state_fidelity(vec_mct, vec_groundtruth)
+        self.assertAlmostEqual(s_f, 1)
+
+    @unpack
+    @data(*itertools.product(
+        [1, 2, 3, 4, 5], ['basic-dirty-ancilla', 'advanced', 'noancilla']
+    ))
+    def test_multi_control_toffoli_matrix_dirty_ancillas(self, num_controls, mode):
```

Comment:
I think we have a `combine` decorator for exactly these situations. I might be wrong (I'm not executing the tests rn)

Reply:
We do, thanks, I wasn't aware of that!

Reply:
Yes we do! Thanks, I wasn't aware of that. Resolved by 90101d5.